### DISCUSSION TITLE
Not thread-safe?

### DISCUSSION THREAD

**Discussion by dca00** [openssl/openssl#24808] [Q&A] [answered]
There is a Windows application in C++ 10 SDK v142 that links to OpenSSL 3.3.1 binary obtained from https://slproweb.com/products/Win32OpenSSL.html
It computes and verifies HMAC SHA256, by a single-step HMAC() call, in multiple threads, in a loop, from several sub-buffers of the main buffer each.
One thread always finds the correct match to a given HMAC.
Sometimes but not always, other threads throw false-positives, i.e. they "think" that they also found a match.

Each thread does something along the lines of:

```
auto i=0;
for(; i<bigBufferSize-31;i++)
{
   auto hmac = HMAC(EVP_sha256(), bigBuffer+i, keylen...);
   if(hmac is the same as known one and global flag is not raised)
   {
     RaiseGlobalFlag_MutexProtected();
     return true; // <= Intermittent false-positives here
   }

    unsigned char *manipulatedSubBuffer;
    Manipulate(bigBuffer+i, manipulatedSubBuffer, keylen);

   hmac = HMAC(EVP_sha256(), manipulatedSubBuffer, keylen...);
   if(hmac is the same as known one and global flag is not raised)
   {
     RaiseGlobalFlag_MutexProtected();
     return true; // <= Intermittent false-positives here
   }

   // and so on, for several more manipulations
}

```
I dumped the memory address and contents of the function's local buffer `hmac` that HMAC() returns, and both the 'right' and the 'wrong' threads show the same address with the same contents.

Does this sound like expected behavior, and am I doing something horribly wrong, or is this a bug that should be filed?

**Comment by t8m** [comment]
HMAC() is thread safe only in case you use non-NULL md parameter for the buffer that should contain the resulting HMAC value. When md is NULL the output will be placed into a static buffer which is obviously not thread safe.

**Reply by dca00** [reply]
I want to ensure that I understand you correctly.
What are you referring to as 'placed into a static buffer'? Doesn't HMAC() return its MD, i.e. places it on the stack?

**Reply by t8m** [reply]
Please show me how do you call the HMAC() function. I.e., the whole call, not just first few arguments.

**Reply by dca00** [reply]
What are you referring to as 'placed into a static buffer'? Doesn't HMAC() return its MD, i.e. places it on the stack?

**Reply by mattcaswell** [reply, marked as answer]
The HMAC function looks like this:

```` C
unsigned char *HMAC(const EVP_MD *evp_md, const void *key, int key_len,
                    const unsigned char *data, size_t data_len,
                    unsigned char *md, unsigned int *md_len)
````
If the `md` argument is `NULL` then the HMAC is stored in a static buffer which is shared between all threads. This is not thread safe. 

From the docs:

> HMAC() computes the message authentication code of the data_len bytes at data using the hash function evp_md and the key key which is key_len bytes long. The key may also be NULL with key_len being 0.
> 
> It places the result in md (which must have space for the output of the hash function, which is no more than EVP_MAX_MD_SIZE bytes). If md is NULL, the digest is placed in a static array. The size of the output is placed in md_len, unless it is NULL. Note: passing a NULL value for md to use the static array is not thread safe.

https://www.openssl.org/docs/man3.3/man3/HMAC.html
